### ISSUE TITLE
t/op/pack.t fails two tests in GCC "long doubles" builds on Windows

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16265]
Migrated from [rt.perl.org#132479](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=132479) (status was 'resolved')

Searchable as RT132479$

**Comment by p5pRT** [comment]
### From @steve-m-hay

#### Created by @steve-m-hay



As noted in https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123113#txn-1317316
there are two test failures when doing a GCC "long doubles" build on Windows&#8203;:

C&#8203;:\\Dev\\Git\\perl\\t>\.\\perl harness op\\pack\.t
op/pack\.t \.\. 1/14713 \# Failed test 13177 \- at op/pack\.t line 1418
\#      got "173 1\.28347651700000004e\-045 42 215 173
1\.28347651700000004e\-045 42215"
\# expected "173 1\.283476517e\-045 42 215 173 1\.283476517e\-045 42 215"
\# Failed test 13180 \- at op/pack\.t line 1431
\#      got "173 1\.28347651700000004e\-045 42 215"
\# expected "173 1\.283476517e\-045 42 215"
op/pack\.t \.\. Failed 2/14713 subtests
&nbsp;    \(less 270 skipped subtests&#8203;: 14441 okay\)

These failures are being recorded here in a separate ticket so that the earlier
one can now be closed\. \(The other outstanding work from it \-\- adding ICC support
for "long doubles" \-\- is also now logged in a separate ticket&#8203;: \#132478\.\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.27.7:

Configured by shay at Tue Nov 21 13:25:42 2017.

Summary of my perl5 (revision 5 version 27 subversion 7) configuration:

  Platform:
    osname=MSWin32
    osvers=6.3.9600
    archname=MSWin32-x86-multi-thread-ld
    uname=''
    config_args='undef'
    hint=recommended
    useposix=true
    d_sigaction=undef
    useithreads=define
    usemultiplicity=define
    use64bitint=undef
    use64bitall=undef
    uselongdouble=define
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='gcc'
    ccflags =' -s -O2 -DWIN32  -DPERL_TEXTMODE_SCRIPTS
-DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -D__USE_MINGW_ANSI_STDIO
-fwrapv -fno-strict-aliasing -mms-bitfields'
    optimize='-s -O2'
    cppflags='-DWIN32'
    ccversion=''
    gccversion='7.1.0'
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=4
    doublesize=8
    byteorder=1234
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=12
    longdblkind=3
    ivtype='long'
    ivsize=4
    nvtype='long double'
    nvsize=12
    Off_t='long long'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='g++'
    ldflags ='-s -L"c:\perl\lib\CORE"
-L"C:\Dev\Software\MinGW-w64\x86\7.1.0\lib"'
    libpth=C:\Dev\Software\MinGW-w64\x86\7.1.0\lib
    libs=-lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool -lcomdlg32
-ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid -lws2_32
-lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    perllibs=-lmoldname -lkernel32 -luser32 -lgdi32 -lwinspool
-lcomdlg32 -ladvapi32 -lshell32 -lole32 -loleaut32 -lnetapi32 -luuid
-lws2_32 -lmpr -lwinmm -lversion -lodbc32 -lodbccp32 -lcomctl32
    libc=
    so=dll
    useshrplib=true
    libperl=libperl527.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags='-mdll -s -L"c:\perl\lib\CORE"
-L"C:\Dev\Software\MinGW-w64\x86\7.1.0\lib"'



@INC for perl 5.27.7:
    C:/perl/site/lib
    C:/perl/lib


Environment for perl 5.27.7:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=C:\Dev\Software\MinGW-w64\x86\7.1.0\bin;C:\Windows\system32;C:\Windows;C:\Windows\system32\wbem;C:\perl\bin
    PERL_BADLANG (unset)
    SHELL (unset)

```
</details>

**Comment by p5pRT** [comment]
### From @sisyphus


> As noted in https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123113#txn-1317316 
> there are two test failures when doing a GCC "long doubles" build on 
> Windows&#8203;:

Just attaching the simple patch to pack\.t that fixes the bug\.

Steve\, could you apply it please ?

\(Then \*both\* tickets can be closed ;\-\)

Cheers\,
Rob

**Comment by p5pRT** [comment]
### From @sisyphus

<details><summary>pack.t.patch</summary>

```diff
--- pack.t_orig	2017-10-21 23:32:24 +1100
+++ pack.t	2017-10-21 23:34:37 +1100
@@ -1414,7 +1414,7 @@
   my @b = unpack "$t X[$t] $t", $p;	# Extract, step back, extract again
   is(scalar @b, 2 * scalar @a);
   $b = "@b";
-  $b =~ s/(?:17000+|16999+)\d+(e-45) /17$1 /gi; # stringification is gamble
+  $b =~ s/(?:17000+|16999+)\d+(e-0?45) /17$1 /gi; # stringification is gamble
   is($b, "@a @a");
 
   use warnings qw(NONFATAL all);;
@@ -1427,7 +1427,7 @@
   is($warning, undef);
   is(scalar @b, scalar @a);
   $b = "@b";
-  $b =~ s/(?:17000+|16999+)\d+(e-45) /17$1 /gi; # stringification is gamble
+  $b =~ s/(?:17000+|16999+)\d+(e-0?45) /17$1 /gi; # stringification is gamble
   is($b, "@a");
 }
```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @steve-m-hay

On Tue\, 21 Nov 2017 13&#8203;:49&#8203;:10 \-0800\, sisyphus wrote&#8203;:
> 
> > As noted in https://rt-archive.perl.org/perl5/Ticket/Display.html?id=123113#txn-1317316 
> > there are two test failures when doing a GCC "long doubles" build on 
> > Windows&#8203;:
> 
> Just attaching the simple patch to pack\.t that fixes the bug\.
> 
> Steve\, could you apply it please ?
> 

Great\! Thanks for the quick fix\! Applied in commit 0e45f4872e73c83376fcd3a255ffcfc823aec531\.

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'resolved'
